### PR TITLE
Fix #518 for 04-pipefilter

### LINKED ISSUE #518: Missing solutions to, context for, and files required by exercises
Many of the exercises don't have solutions. Some don't need solutions because learners can run the commands and see which is the correct answer. In some of the exercises, the commands can't be tested because the data files don't exist e.g. 07-find, "matching and subtracting" - there are no *ose.dat files

> The -v flag to grep inverts pattern matching, so that only lines which do not match the pattern are printed. Given that, which of the following commands will find all files in /data whose names end in ose.dat (e.g., sucrose.dat or maltose.dat), but do not contain the word temp?
> 
> find /data -name '*.dat' | grep ose | grep -v temp
> find /data -name ose.dat | grep -v temp
> grep -v "temp" $(find /data -name '*ose.dat')
> None of the above.

Sometimes it isn't clear which directory the question refers to.

I plan to submit a series of PRs to address these issues, using existing files as much as possible. Is this of interest to others?

### PR DESCRIPTION
**Pull request by gcapes**
Add solutions, provide working directory context, and where necessary modify the exercises to work with files in data-shell (so that learners can test commands rather than having to work everything out in their heads).

### CODE REVIEW

File: _episodes/04-pipefilter.md [hunk 1]
```diff
@@ -522,23 +522,60 @@ so this matches all the valid data files she has.
...
+> > ## Solution
+> > The `-n` flag specifies a numeric sort, rather than alphabetical.
+> {: .solution}
 {: .challenge}
 
 > ## What Does `<` Mean?
 >
+> Change directory to data-shell (the top level of our downloaded example data)
+> ```
+> cd ~/data-shell
```
Comment (shwina) [inline _episodes/04-pipefilter.md]: Remove this `cd` command, as people may have `data-shell` in e.g., their Desktop

File: _episodes/04-pipefilter.md [hunk 2]
```diff
@@ -522,23 +522,60 @@ so this matches all the valid data files she has.
...
+> $ wc -l < notes.txt
 > ~~~
 > {: .bash}
+>
+> > ## Solution
+> > `<` is used to redirect input to a command. 
+> >
+> > In both examples, the shell returns the number of lines from the input to
+> > the `wc` command.
+> > In the first example, the input is a the file notes.txt and the file name is
```
Comment (shwina) [inline _episodes/04-pipefilter.md]: backticks: `notes.txt`

File: _episodes/04-pipefilter.md [hunk 3]
```diff
@@ -605,12 +651,20 @@ so this matches all the valid data files she has.
 > 2. `wc -l * | sort -n | head -n 1-3`
 > 3. `wc -l * | head -n 3 | sort -n`
 > 4. `wc -l * | sort -n | head -n 3`
+>
+> > ## Solution
+> > Option 4 is the solution.
+> > The pipe character `|` is used to feed the standard output from one process to
+> > the standard input of another.
+> > `>` is used to redirect standard output to a file.
+> > Try it in the molecules directory!
```
Comment (shwina) [inline _episodes/04-pipefilter.md]: Backticks: `molecules`

Comment (shwina) [inline _episodes/04-pipefilter.md]: Also, perhaps provide the full path to `molecules`, starting from `data-shell` as the root

File: _episodes/04-pipefilter.md [hunk 4]
```diff
@@ -605,12 +651,20 @@ so this matches all the valid data files she has.
...
+> {: .solution}
 {: .challenge}
 
 > ## Why Does `uniq` Only Remove Adjacent Duplicates?
 >
 > The command `uniq` removes adjacent duplicated lines from its input.
-> For example, if a file `salmon.txt` contains:
+> For example, the file data-shell/data/salmon.txt contains:
```
Comment (shwina) [inline _episodes/04-pipefilter.md]: Backticks

File: _episodes/04-pipefilter.md [hunk 5]
```diff
@@ -635,11 +689,18 @@ so this matches all the valid data files she has.
...
+> > $ sort salmon.txt | uniq
+> > ```
+> > {: .bash}
+> {: .solution}
 {: .challenge}
 
 > ## Pipe Reading Comprehension
 >
-> A file called `animals.txt` contains the following data:
+> A file called `animals.txt` (in the data-shell/data folder) contains the following data:
```
Comment (shwina) [inline _episodes/04-pipefilter.md]: Backticks

File: _episodes/04-pipefilter.md [hunk 6]
```diff
@@ -659,6 +720,7 @@ so this matches all the valid data files she has.
 > $ cat animals.txt | head -n 5 | tail -n 3 | sort -r > final.txt
 > ~~~
 > {: .bash}
+> Hint: build the pipe line up one command at a time to test your understanding
```
Comment (shwina) [inline _episodes/04-pipefilter.md]: pipe line -> pipeline?

File: _episodes/04-pipefilter.md [hunk 7]
```diff
@@ -701,13 +770,25 @@ so this matches all the valid data files she has.
 > 2. `rm *.txt`
 > 3. `rm * .txt`
 > 4. `rm *.*`
+>
+> > ## Solution
+> > 1. This would remove .txt files with one-character names
+> > 2. This is correct answer
+> > 3. The shell would expand * to match everything in the current directory,
```
Comment (shwina) [inline _episodes/04-pipefilter.md]: Backticks for `*`

File: _episodes/04-pipefilter.md [hunk 8]
```diff
@@ -701,13 +770,25 @@ so this matches all the valid data files she has.
...
+> > so the command would try to remove all matched files and an additional
+> > file called .txt
+> > 4. The shell would expand *.* to match all files with any extension,
```
Comment (shwina) [inline _episodes/04-pipefilter.md]: Backticks for `.`

File: _episodes/04-pipefilter.md [hunk 9]
```diff
@@ -701,13 +770,25 @@ so this matches all the valid data files she has.
 > 2. `rm *.txt`
 > 3. `rm * .txt`
 > 4. `rm *.*`
+>
+> > ## Solution
+> > 1. This would remove .txt files with one-character names
+> > 2. This is correct answer
+> > 3. The shell would expand * to match everything in the current directory,
+> > so the command would try to remove all matched files and an additional
+> > file called .txt
```
Comment (shwina) [inline _episodes/04-pipefilter.md]: Backticks: `.txt`

File: _episodes/04-pipefilter.md [hunk 10]
```diff
@@ -701,13 +770,25 @@ so this matches all the valid data files she has.
...
+> {: .solution}
 {: .challenge}
 
 > ## Wildcard Expressions
 >
 > Wildcard expressions can be very complex, but you can sometimes write
 > them in ways that only use simple syntax, at the expense of being a bit
-> more verbose.  For example, the wildcard expression `*[AB].txt`
+> more verbose.  
+> Consider the directory data-shell/north-pacific-gyre/2012-07-03 :
```
Comment (shwina) [inline _episodes/04-pipefilter.md]: Backticks

File: _episodes/04-pipefilter.md [hunk 11]
```diff
@@ -721,11 +802,24 @@ so this matches all the valid data files she has.
...
+> > 2. The output from the new commands is separated because there are two commands.
+> > 3. When there are no files ending in `A.txt`, or there are no files ending in
+> > `B.txt`.
+> {: .solution}
 {: .challenge}
 
 > ## Which Pipe?
 >
-> A file called animals.txt contains 586 lines of data formatted as follows:
+> The file data-shell/data/animals.txt contains 586 lines of data formatted as follows:
```
Comment (shwina) [inline _episodes/04-pipefilter.md]: Backticks

File: _episodes/04-pipefilter.md [hunk 12]
```diff
@@ -736,7 +830,8 @@ so this matches all the valid data files she has.
 > ~~~
 > {: .output}
 >
-> What command would you use to produce a table that shows
+> Assuming your current directory is data-shell/data/,
```
Comment (shwina) [inline _episodes/04-pipefilter.md]: Backticks

File: _episodes/04-pipefilter.md [hunk 13]
```diff
@@ -745,13 +840,19 @@ so this matches all the valid data files she has.
 > 4.  `cut -d, -f 2 animals.txt | uniq -c`
 > 5.  `cut -d, -f 2 animals.txt | sort | uniq -c`
 > 6.  `cut -d, -f 2 animals.txt | sort | uniq -c | wc -l`
+>
+> > ## Solution
+> > 5. is the correct answer.
+> > If you have difficulty understanding why, try running the commands, or sub-sections of
+> > the pipe lines (make sure you are in the data-shell/data directory).
```
Comment (shwina) [inline _episodes/04-pipefilter.md]: 1. pipe lines -> pipelines
2. Backticks

**Comment by gcapes** [comment]
Hi @shwina, 
Are there guidelines somewhere for when SWC use the single backticks? I thought these were for inline code/commands, but it looks we're using them for anything that might be typed into a terminal? Rebase to follow shortly.
Thanks
G

**Comment by rgaiacs** [comment]
> Are there guidelines somewhere for when SWC use the single backticks?

No. :-( Our style guide is pretty bad at the moment.

> looks we're using them for anything that might be typed into a terminal

More or less.

**Comment by gcapes** [comment]
Ok, thanks!

**Comment by shwina** [comment]
Thanks, @gcapes !

**Pull request merged**